sdk/core/core-rest-pipeline/src/fetchHttpClient.ts

```diff
@@ -0,0 +1,237 @@
+    response.bodyAsText = await responseStream.text();
+  }
+
+  return response;
+}
+
+function handleAbortSignal(request: PipelineRequest): AbortController {
+  const abortController = new AbortController();
+  /**
+   * Attach an abort listener to the request
+   */
+  let abortListener: ((event: any) => void) | undefined;
+  if (request.abortSignal) {
+    if (request.abortSignal.aborted) {
+      throw new AbortError("The operation was aborted.");
+    }
+
+    abortListener = (event: Event) => {
+      if (event.type === "abort") {
+        abortController.abort();
+      }
+    };
+    request.abortSignal.addEventListener("abort", abortListener);
+  }
+
+  // If a timeout was passed, call the abort signal once the time elapses
+  if (request.timeout > 0) {
+    setTimeout(() => {
+      abortController.abort();
+    }, request.timeout);
+  }
+
+  return abortController;
+}
+
+/**
+ * Gets the specific error
+ */
+function getError(e: RestError, request: PipelineRequest): RestError {
+  if (e && e?.name === "AbortError") {
+    return e;
+  } else {
+    return new RestError(`Error sending request: ${e.message}`, {
+      code: e?.code ?? RestError.REQUEST_SEND_ERROR,
+      request,
+    });
+  }
+}
+
+/**
+ * Converts PipelineRequest headers to Fetch headers
+ */
+function buildFetchHeaders(pipelineHeaders: PipelineHeaders) {
+  const headers = new Headers();
+  for (const [name, value] of pipelineHeaders) {
+    headers.append(name, value);
+  }
+
+  return headers;
+}
+
+function buildPipelineHeaders(httpResponse: Response): PipelineHeaders {
+  const responseHeaders = createHttpHeaders();
+  httpResponse.headers.forEach((value, key) => {
+    responseHeaders.set(key, value);
+  });
+
+  return responseHeaders;
+}
+
+function buildRequestBody(request: PipelineRequest) {
+  if (isNodeReadableStream(request.body)) {
+    throw new Error("Node streams are not supported in browser environment.");
+  }
+
+  return isReadableStream(request.body)
+    ? buildBodyStream(request.body, request.onUploadProgress)
+    : request.body;
+}
+
+/**
+ * Reads the request/response original stream and stream it through a new
+ * ReadableStream, this is done to be able to report progress in a way that
+ * all modern browsers support. TransformStreams would be an alternative,
+ * however they are not yet supported by all browsers i.e Firefox
+ */
+function buildBodyStream(
+  readableStream: ReadableStream<Uint8Array>,
+  onProgress?: (progress: TransferProgressEvent) => void
+): ReadableStream<Uint8Array> {
+  const reader = readableStream.getReader();
+
+  return new ReadableStream({
+    async start(controller) {
+      // eslint-disable-next-line no-constant-condition
+      while (true) {
+        const { done, value } = await reader.read();
+
+        // When no more data needs to be consumed, break the reading
+        if (done) {
+          break;
+        }
+
+        if (onProgress) {
+          onProgress({ loadedBytes: value?.length ?? 0 });
```

Comment:
don't we need to be keeping count of how many bytes were in previous chunks? I think this number is supposed to be cumulative.